ext/network.go

```diff
@@ -0,0 +1,540 @@
+	}
+	return types.Bool(net.ParseIP(string(s)) != nil)
+}
+
+func netIsCIDR(val ref.Val) ref.Val {
+	s, ok := val.(types.String)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(val)
+	}
+	_, _, err := net.ParseCIDR(string(s))
+	return types.Bool(err == nil)
+}
+
+func netIPString(val ref.Val) ref.Val {
+	s, ok := val.(types.String)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(val)
+	}
+	str := string(s)
+	ip := net.ParseIP(str)
+	if ip == nil {
+		return types.NewErr("invalid ip address: %s", str)
+	}
+	return IP{IP: ip, Str: str}
+}
+
+func netCIDRString(val ref.Val) ref.Val {
+	s, ok := val.(types.String)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(val)
+	}
+	str := string(s)
+	ip, ipNet, err := net.ParseCIDR(str)
+	if err != nil {
+		return types.NewErr("invalid cidr range: %s", str)
+	}
+	// Store the specific IP (which might have host bits set) alongside the network
+	return CIDR{IPNet: ipNet, ExtraIP: ip, Str: str}
+}
+
+func netIPFamily(val ref.Val) ref.Val {
+	ip, ok := val.(IP)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(val)
+	}
+	if ip.IP.To4() != nil {
+		return types.Int(4)
+	}
+	return types.Int(6)
+}
+
+func netIPIsCanonical(val ref.Val) ref.Val {
+	ip, ok := val.(IP)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(val)
+	}
+	return types.Bool(ip.Str == ip.IP.String())
+}
+
+func netIPIsLoopback(val ref.Val) ref.Val {
+	ip, ok := val.(IP)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(val)
+	}
+	return types.Bool(ip.IP.IsLoopback())
+}
+
+func netIPIsGlobalUnicast(val ref.Val) ref.Val {
+	ip, ok := val.(IP)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(val)
+	}
+	return types.Bool(ip.IP.IsGlobalUnicast())
+}
+
+func netIPIsUnspecified(val ref.Val) ref.Val {
+	ip, ok := val.(IP)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(val)
+	}
+	return types.Bool(ip.IP.IsUnspecified())
+}
+
+func netIPIsLinkLocalMulticast(val ref.Val) ref.Val {
+	ip, ok := val.(IP)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(val)
+	}
+	return types.Bool(ip.IP.IsLinkLocalMulticast())
+}
+
+func netIPIsLinkLocalUnicast(val ref.Val) ref.Val {
+	ip, ok := val.(IP)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(val)
+	}
+	return types.Bool(ip.IP.IsLinkLocalUnicast())
+}
+
+func netCIDRContainsIP(lhs, rhs ref.Val) ref.Val {
+	cidr, ok := lhs.(CIDR)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(lhs)
+	}
+	ip, ok := rhs.(IP)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(rhs)
+	}
+	return types.Bool(cidr.IPNet.Contains(ip.IP))
+}
+
+func netCIDRContainsIPString(lhs, rhs ref.Val) ref.Val {
+	cidr, ok := lhs.(CIDR)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(lhs)
+	}
+	s, ok := rhs.(types.String)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(rhs)
+	}
+	ip := net.ParseIP(string(s))
+	if ip == nil {
+		return types.NewErr("invalid ip address: %s", s)
+	}
+	return types.Bool(cidr.IPNet.Contains(ip))
+}
+
+func netCIDRContainsCIDR(lhs, rhs ref.Val) ref.Val {
+	parent, ok := lhs.(CIDR)
```

Comment:
These sorts of type guards can be removed since there are generic type guards generated during the function declaration

It should be safe to just write `parent := lhs.(CIDR)`